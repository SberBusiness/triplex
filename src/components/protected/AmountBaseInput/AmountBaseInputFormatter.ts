import {AmountConst} from '@sberbusiness/triplex/consts/AmountConst';

/** Форматер значения в AmountBaseInput. */
export class AmountBaseInputFormatter {
    /** Значение. */
    private value: string;
    /** Величина сдвига каретки. */
    private caretOffset: number;
    /** Максимальное количество знаков (без учёта пробелов). */
    private readonly maxLength: number;
    /** Количество чисел после запятой. */
    private readonly fractionDigits: number;

    constructor(maxLength: number, fractionDigits: number) {
        this.value = '';
        this.caretOffset = 0;
        this.maxLength = maxLength;
        this.fractionDigits = fractionDigits;
    }

    /** Получение значения. */
    public getValue(): string {
        return this.value;
    }

    /** Получение величины сдвига каретки. */
    public getCaretOffset(): number {
        return this.caretOffset;
    }

    /** Применение входных данные для форматирования. */
    public apply(value: string): void {
        const length = value.length;

        if (this.fractionDigits == 0) {
            this.value = this.formatInteger(value, length);
        } else {
            this.value = this.formatDecimal(value, length);
        }
    }

    /** Форматирование значения в виде целого значения. */
    private formatInteger(value: string, length: number): string {
        const buffer: string[] = [];

        if (length > 0) {
            let i = 0;

            buffer.push(value[i]);
            while (++i < length) {
                // Нужно ли добавить разделитель между группами цифр.
                if ((length - i) % 3 == 0) {
                    buffer.push(' ');
                    this.caretOffset++;
                }
                buffer.push(value[i]);
            }
        }

        return buffer.join('');
    }

    /** Форматирование значения в виде десятичной дроби. */
    private formatDecimal(value: string, length: number): string {
        const buffer: string[] = [];

        if (length) {
            const integerLength = length - this.fractionDigits - 1;
            let i = 0;

            buffer.push(value[i]);
            while (++i < integerLength) {
                // Нужно ли добавить разделитель между группами цифр.
                if ((integerLength - i) % 3 == 0) {
                    buffer.push(' ');
                    this.caretOffset++;
                }
                buffer.push(value[i]);
            }

            buffer.push(AmountConst.DecimalComma);

            while (++i < length) {
                buffer.push(value[i]);
            }
        }

        return buffer.join('');
    }
}
