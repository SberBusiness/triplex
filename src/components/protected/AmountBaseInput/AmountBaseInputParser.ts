import {AmountConst} from '@sberbusiness/triplex/consts/AmountConst';
import {StringUtils} from '@sberbusiness/triplex/utils/stringUtils';
import {isKey} from '@sberbusiness/triplex/utils/keyboard';

/** Обработчик значения в AmountBaseInput. */
export class AmountBaseInputParser {
    /** Значение. */
    private value: string;
    /** Величина сдвига каретки. */
    private caretOffset: number;
    /** Значение нажатой клавиши. */
    private key: string;
    /** Максимальное количество знаков перед запятой. */
    private readonly maxIntegerDigits: number;
    /** Количество чисел после запятой. */
    private readonly fractionDigits: number;

    constructor(maxIntegerDigits: number, fractionDigits: number) {
        this.value = '';
        this.caretOffset = 0;
        this.maxIntegerDigits = maxIntegerDigits;
        this.fractionDigits = fractionDigits;
        this.key = '';
    }

    /** Получение значения. */
    public getValue(): string {
        return this.value;
    }

    /** Получение величины сдвига каретки. */
    public getCaretOffset(): number {
        return this.caretOffset;
    }

    /** Применение входных данных для обработки. */
    public apply(value: string, caret: number, key: string): void {
        const length = value.length;

        this.key = key;

        if (this.maxIntegerDigits > 0) {
            if (this.fractionDigits == 0) {
                return this.parseInteger(value, caret, length);
            } else if (this.fractionDigits > 0) {
                return this.parseDecimal(value, caret, length);
            }
        }

        // Fallback
        this.value = '';
        this.caretOffset -= length;
    }

    /** Обработка значения в виде целого числа. */
    private parseInteger(value: string, caret: number, length: number): void {
        this.parseIntegerPart(value, caret, length);
    }

    /** Обработка значения в виде десятичной дроби. */
    private parseDecimal(value: string, caret: number, length: number): void {
        const separatorIndex = this.findSeparatorIndex(value, caret, length);
        const [integerEnd, fractionalStart] = separatorIndex != -1 ? [separatorIndex, separatorIndex + 1] : [caret, caret];
        const [integerPart, fractionalPart] = [value.substring(0, integerEnd), value.substring(fractionalStart)];

        this.parseIntegerPart(integerPart, Math.min(caret, integerEnd), integerEnd);
        this.parseFractionalPart(fractionalPart, Math.max(caret - fractionalStart, 0), length - fractionalStart);
    }

    /** Обработка целой части числа. */
    private parseIntegerPart(value: string, caret: number, length: number): void {
        const buffer: string[] = [];
        [value, caret, length] = this.trimLeadingZeros(value, caret, length);

        // Обработка значения перед кареткой.
        if (caret > 0) {
            const maxDigitsBeforeCaret = this.maxIntegerDigits - (length - caret - Math.floor((length - caret) / 4));

            if (maxDigitsBeforeCaret > 0) {
                for (let i = 0, n = 0; i < caret; i++) {
                    if (StringUtils.isDigit(value[i])) {
                        buffer.push(value[i]);
                        if (++n == maxDigitsBeforeCaret) {
                            break;
                        }
                    }
                }
            }

            this.caretOffset -= caret - buffer.length;
        }

        // Обработка значения после каретки.
        if (caret < length) {
            for (let i = caret; i < length; i++) {
                if (StringUtils.isDigit(value[i])) {
                    buffer.push(value[i]);
                } else {
                    this.caretOffset--;
                }
            }
        }

        this.value = buffer.join('');
    }

    /** Обработка дробной части числа. */
    private parseFractionalPart(value: string, caret: number, length: number): void {
        const buffer: string[] = [];

        // Обработка значения перед кареткой.
        if (caret > 0) {
            const maxDigitsBeforeCaret = this.fractionDigits;

            for (let i = 0, n = 0; i < caret; i++) {
                if (StringUtils.isDigit(value[i])) {
                    buffer.push(value[i]);
                    if (++n == maxDigitsBeforeCaret) {
                        break;
                    }
                } else {
                    this.caretOffset--;
                }
            }
        }

        // Обработка значения после каретки.
        while (buffer.length < this.fractionDigits) {
            const index: number = length - (this.fractionDigits - buffer.length);

            if (index >= caret) {
                buffer.push(value[index]);
            } else {
                buffer.push('0');
                if (isKey(this.key, 'DELETE')) {
                    this.caretOffset++;
                }
            }
        }

        // Если целая часть отсутствует.
        if (this.value.length == 0) {
            // При ненулевой дробной части, восстанавливаем целый ноль.
            if (buffer.some((digit) => digit != '0')) {
                this.value = '0';
                this.caretOffset++;
            }
            // Останавливаем парсинг.
            else {
                return;
            }
        }

        this.value += AmountConst.DecimalPoint;
        this.value += buffer.join('');
    }

    /** Удаление ведущих нулей и всего, что им предшествует. */
    private trimLeadingZeros(value: string, caret: number, length: number): [string, number, number] {
        let start = 0;

        for (let i = 0; i < length; i++) {
            if (StringUtils.isDigit(value[i])) {
                start = i;
                if (value[i] != '0') {
                    break;
                }
            }
        }

        this.caretOffset -= start;

        return [value.substring(start), Math.max(caret - start, 0), length - start];
    }

    /** Нахождение индекса десятичного разделителя. */
    private findSeparatorIndex(value: string, caret: number, length: number): number {
        // Если каретка находится перед числом целой части.
        if (caret < length - (this.fractionDigits + 1)) {
            for (let i = length - 1; i >= 0; i--) {
                if (StringUtils.isDecimalSeparator(value[i])) {
                    return i;
                }
            }
        }
        // Если каретка находится после целой части.
        else {
            for (let i = 0; i < length; i++) {
                if (StringUtils.isDecimalSeparator(value[i])) {
                    return i;
                }
            }
        }

        return -1;
    }
}
