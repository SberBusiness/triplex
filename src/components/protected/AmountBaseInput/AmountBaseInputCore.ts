import {AmountBaseInputParser} from './AmountBaseInputParser';
import {AmountBaseInputFormatter} from './AmountBaseInputFormatter';

/** Свойства кэша AmountBaseInputCore. */
interface IAmountBaseInputCoreCache {
    /** Отформатированное значение. */
    formattedValue: string;
    /** Значение нажатой клавиши. */
    key: string;
    /** Позиция начала выделения. */
    selectionStart: number | null;
    /** Позиция конца выделения. */
    selectionEnd: number | null;
    /** Направление выделения. */
    selectionDirection: 'forward' | 'backward' | 'none' | null;
}

/** Основная логика в AmountBaseInput. */
export class AmountBaseInputCore {
    /** Значение. */
    public value: string;
    /** Отформатированное значение. */
    public formattedValue: string;
    /** Положение каретки. */
    public caret: number;
    /** Максимальное количество знаков перед запятой. */
    public maxIntegerDigits: number;
    /** Количество чисел после запятой. */
    public fractionDigits: number;
    /** Кэш для хранения значений. */
    public cache: IAmountBaseInputCoreCache;

    constructor(maxIntegerDigits: number, fractionDigits: number) {
        this.value = '';
        this.formattedValue = '';
        this.caret = 0;
        this.maxIntegerDigits = maxIntegerDigits;
        this.fractionDigits = fractionDigits;
        this.cache = {formattedValue: '', key: '', selectionDirection: null, selectionEnd: null, selectionStart: null};
    }

    /** Применение входных данных. */
    public apply(value: string, caret: number): void {
        this.parse(value, caret);
        this.format(this.value);
    }

    /** Обработка значения. */
    private parse(value: string, caret: number): void {
        const parser = new AmountBaseInputParser(this.maxIntegerDigits, this.fractionDigits);

        parser.apply(value, caret, this.cache.key);

        this.value = parser.getValue();
        this.caret = caret + parser.getCaretOffset();
    }

    /** Форматирование значения. */
    private format(value: string): void {
        const formatter = new AmountBaseInputFormatter(this.maxIntegerDigits, this.fractionDigits);

        formatter.apply(value);

        this.formattedValue = formatter.getValue();
        this.caret += formatter.getCaretOffset();
    }
}
