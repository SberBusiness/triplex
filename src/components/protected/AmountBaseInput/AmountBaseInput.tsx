import React, {useRef, useMemo, useLayoutEffect} from 'react';
import {IInputProps, Input} from '@sberbusiness/triplex/components/Input/Input';
import {setCaretPosition} from '@sberbusiness/triplex/utils/inputUtils';
import {isKey} from '@sberbusiness/triplex/utils/keyboard';
import {AmountConst} from '@sberbusiness/triplex/consts/AmountConst';
import {StringUtils} from '@sberbusiness/triplex/utils/stringUtils';
import {AmountBaseInputCore} from './AmountBaseInputCore';

/** Свойства компонента AmountBaseInput. */
interface IAmountBaseInputProps extends Omit<IInputProps, 'type' | 'maxLength' | 'onChange'> {
    /** Значение. */
    value: string;
    /** Максимальное количество знаков перед запятой. */
    maxIntegerDigits?: number;
    /** Количество знаков после запятой. */
    fractionDigits?: number;
    /** Обработчик изменения значения. */
    onChange: (value: string) => void;
}

/** База для поля ввода суммы. */
export const AmountBaseInput = React.forwardRef<HTMLInputElement, IAmountBaseInputProps>(
    ({value, placeholder, onKeyDown, onSelect, maxIntegerDigits = 16, fractionDigits = 2, onChange, ...rest}, ref) => {
        const core = useRef(new AmountBaseInputCore(maxIntegerDigits, fractionDigits));
        const inputRef = useRef<HTMLInputElement | null>(null);

        useLayoutEffect(() => {
            if (inputRef.current == document.activeElement) {
                setCaretPosition(inputRef.current, Math.max(core.current.caret, 0));
            }
        }, [value]);

        /** Функция, возвращающая отформатированное значение. */
        const getFormattedValue = () => {
            if (
                value != core.current.value ||
                maxIntegerDigits != core.current.maxIntegerDigits ||
                fractionDigits != core.current.fractionDigits
            ) {
                core.current.maxIntegerDigits = maxIntegerDigits;
                core.current.fractionDigits = fractionDigits;
                core.current.apply(value, value.length);
            }

            core.current.cache.formattedValue = core.current.formattedValue;

            return core.current.formattedValue;
        };

        /** Текст подсказки по-умолчанию. */
        const defaultPlaceholder = useMemo(() => {
            const buffer: string[] = [];

            buffer.push('0');
            if (fractionDigits > 0) {
                buffer.push(AmountConst.DecimalComma);
                for (let i = 0; i < fractionDigits; i++) {
                    buffer.push('0');
                }
            }

            return buffer.join('');
        }, [fractionDigits]);

        /** Обработчик изменения значения. */
        const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
            const caret = event.target.selectionStart ?? event.target.value.length;

            core.current.apply(event.target.value, caret);

            setFallbackCaret(event.target);

            onChange(core.current.value);
        };

        /**
         * Установка каретки на случай, если не произойдёт изменения значения.
         *
         * Значение может не измениться по двум причинам:
         *     1) Значение невалидно. Обработчик снаружи не обновляет состояние;
         *     2) Значение идентично предыдущему. Обработчик снаружи [не] обновляет состояние.
         **/
        const setFallbackCaret = (input: HTMLInputElement) => {
            const {formattedValue, key, selectionStart, selectionEnd, selectionDirection} = core.current.cache;

            input.value = formattedValue;

            if (selectionStart == null || selectionEnd == null || selectionDirection == null) {
                return;
            }

            if (key == AmountConst.DecimalComma || key == AmountConst.DecimalPoint) {
                // Если каретка находится непосредственно перед десятичным разделителем, сдвигаем каретку вперёд.
                if (selectionStart == selectionEnd && formattedValue[selectionStart] == AmountConst.DecimalComma) {
                    return input.setSelectionRange(selectionStart + 1, selectionStart + 1);
                }
            } else if (isKey(key, 'BACKSPACE')) {
                if (selectionStart == selectionEnd) {
                    // Если каретка стоит непосредственно после группового/десятичного разделителя, сдвигаем каретку назад.
                    if (StringUtils.isDigit(formattedValue[selectionStart - 1]) == false) {
                        return input.setSelectionRange(selectionStart - 1, selectionEnd - 1);
                    }
                }
                // Если есть выделенный текст, ставим каретку в начало выделения.
                else {
                    // Ставим каретку в начало выделения.
                    return input.setSelectionRange(selectionStart, selectionStart);
                }
            } else if (isKey(key, 'DELETE')) {
                if (selectionStart == selectionEnd) {
                    // Если каретка стоит непосредственно перед групповым/десятичным разделителем, сдвигаем каретку вперёд.
                    if (StringUtils.isDigit(formattedValue[selectionStart]) == false) {
                        return input.setSelectionRange(selectionStart + 1, selectionEnd + 1);
                    }
                }
                // Если есть выделенный текст, ставим каретку в конец выделения.
                else {
                    return input.setSelectionRange(selectionEnd, selectionEnd);
                }
            }

            // Если каретка находится после десятичного разделителя.
            if (fractionDigits > 0 && formattedValue.length - fractionDigits - 1 < selectionStart) {
                // Если новое значение повторяет старое, ставим каретку в расчётное место.
                if (core.current.formattedValue == formattedValue) {
                    return input.setSelectionRange(core.current.caret, core.current.caret);
                }
            }

            // Если текст выделялся в обратном порядке, ставим каретку в конец выделения.
            if (selectionDirection == 'backward') {
                return input.setSelectionRange(selectionEnd, selectionEnd);
            }

            // В остальных случаях ставим каретку в начало выделения.
            return input.setSelectionRange(selectionStart, selectionStart);
        };

        /** Обработчик нажатия клавиши. */
        const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
            core.current.cache.key = event.key;
            onKeyDown?.(event);
        };

        /** Обработчик выбора текста. */
        const handleSelect = (event: React.SyntheticEvent<HTMLInputElement>) => {
            core.current.cache.key = '';
            core.current.cache.selectionStart = event.currentTarget.selectionStart;
            core.current.cache.selectionEnd = event.currentTarget.selectionEnd;
            core.current.cache.selectionDirection = event.currentTarget.selectionDirection;
            onSelect?.(event);
        };

        /** Функция для хранения ссылки. */
        const setRef = (instance: HTMLInputElement | null) => {
            inputRef.current = instance;
            if (typeof ref === 'function') {
                ref(instance);
            } else if (ref) {
                ref.current = instance;
            }
        };

        return (
            <Input
                value={getFormattedValue()}
                placeholder={placeholder || defaultPlaceholder}
                autoComplete="off"
                onChange={handleChange}
                onKeyDown={handleKeyDown}
                onSelect={handleSelect}
                {...rest}
                ref={setRef}
            />
        );
    }
);

AmountBaseInput.displayName = 'AmountBaseInput';
