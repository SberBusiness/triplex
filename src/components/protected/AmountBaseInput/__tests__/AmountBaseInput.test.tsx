import React from 'react';
import renderer from 'react-test-renderer';
import {AmountBaseInput} from '../AmountBaseInput';

describe('AmountBaseInput', () => {
    beforeEach(() => {
        allure.feature('AmountBaseInput');
    });

    it('renders correctly', () => {
        const component = <AmountBaseInput value="1234567890" onChange={() => {}} />;
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
