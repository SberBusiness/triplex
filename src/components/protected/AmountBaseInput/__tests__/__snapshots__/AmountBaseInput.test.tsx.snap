// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountBaseInput renders correctly 1`] = `
<input
  autoComplete="off"
  className="cssClass[input]"
  onChange={[Function]}
  onKeyDown={[Function]}
  onSelect={[Function]}
  placeholder="0,00"
  value="1 234 567 890,00"
/>
`;
